File: Windows Source/TotallyHuman/Utilities/AudioFileHelper.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using TotallyHuman.Models;

namespace TotallyHuman.Utilities;

/// <summary>
/// Hilfsfunktionen rund um Audiodateien: unterstützte Formate, MD5-Signatur
/// (für Deduplizierung) und rekursives Auffinden von Audiodateien. Port von
/// AudioFileHelper.swift.
/// </summary>
public static class AudioFileHelper
{
    public static readonly HashSet<string> UnterstuetzteEndungen = AnalysisConfig.SupportedExtensions;

    public static bool IstAudioDatei(string pfad)
        => UnterstuetzteEndungen.Contains(Path.GetExtension(pfad).TrimStart('.').ToLowerInvariant());

    /// <summary>MD5-Signatur des Dateiinhalts (streaming, speicherschonend).</summary>
    public static string Md5Signatur(string pfad)
    {
        try
        {
            using var stream = File.OpenRead(pfad);
            using var md5 = MD5.Create();
            byte[] hash = md5.ComputeHash(stream);
            return ToHex(hash);
        }
        catch
        {
            // Fallback: Pfad-basierter Hash
            using var md5 = MD5.Create();
            return ToHex(md5.ComputeHash(Encoding.UTF8.GetBytes(pfad)));
        }
    }

    private static string ToHex(byte[] bytes)
    {
        var sb = new StringBuilder(bytes.Length * 2);
        foreach (var b in bytes) sb.Append(b.ToString("x2"));
        return sb.ToString();
    }

    /// <summary>Findet alle unterstützten Audiodateien (rekursiv), oder gibt die Datei selbst zurück.</summary>
    public static List<string> FindeAudioDateien(string pfad)
    {
        var ergebnis = new List<string>();
        if (File.Exists(pfad))
        {
            if (IstAudioDatei(pfad)) ergebnis.Add(pfad);
            return ergebnis;
        }
        if (!Directory.Exists(pfad)) return ergebnis;

        try
        {
            foreach (var f in Directory.EnumerateFiles(pfad, "*", SearchOption.AllDirectories))
            {
                var name = Path.GetFileName(f);
                if (name.StartsWith(".")) continue; // versteckte Dateien überspringen
                if (IstAudioDatei(f)) ergebnis.Add(f);
            }
        }
        catch { /* Zugriffsfehler ignorieren */ }

        return ergebnis.OrderBy(p => Path.GetFileName(p), StringComparer.CurrentCulture).ToList();
    }
}

File: macOS Source/TotallyHuman/Utilities/AudioFileHelper.swift
import Foundation
import CryptoKit

/// Hilfsfunktionen rund um Audiodateien: unterstützte Formate, Datei-Signatur
/// (für Deduplizierung) und rekursives Auffinden von Audiodateien in Ordnern.
enum AudioFileHelper {

    static let unterstuetzteEndungen: Set<String> = AnalysisConfig.supportedExtensions

    static func istAudioDatei(_ url: URL) -> Bool {
        unterstuetzteEndungen.contains(url.pathExtension.lowercased())
    }

    /// MD5-Signatur des Dateiinhalts (streaming, speicherschonend).
    static func md5Signatur(url: URL) -> String {
        guard let stream = InputStream(url: url) else {
            // Fallback: Pfad + Größe
            return md5(of: Data(url.path.utf8))
        }
        stream.open()
        defer { stream.close() }
        var hasher = Insecure.MD5()
        let puffer = UnsafeMutablePointer<UInt8>.allocate(capacity: 65536)
        defer { puffer.deallocate() }
        while stream.hasBytesAvailable {
            let gelesen = stream.read(puffer, maxLength: 65536)
            if gelesen <= 0 { break }
            hasher.update(data: Data(bytes: puffer, count: gelesen))
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    private static func md5(of data: Data) -> String {
        Insecure.MD5.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    /// Findet alle unterstützten Audiodateien in einem Ordner (rekursiv). Ist die
    /// URL selbst eine Datei, wird sie – falls unterstützt – zurückgegeben.
    static func findeAudioDateien(in url: URL) -> [URL] {
        let fm = FileManager.default
        var istOrdner: ObjCBool = false
        guard fm.fileExists(atPath: url.path, isDirectory: &istOrdner) else { return [] }
        if !istOrdner.boolValue {
            return istAudioDatei(url) ? [url] : []
        }
        var ergebnis: [URL] = []
        if let e = fm.enumerator(at: url, includingPropertiesForKeys: [.isRegularFileKey], options: [.skipsHiddenFiles]) {
            for case let f as URL in e where istAudioDatei(f) {
                ergebnis.append(f)
            }
        }
        return ergebnis.sorted { $0.lastPathComponent.localizedCompare($1.lastPathComponent) == .orderedAscending }
    }
}
